### PR TITLE
[SPARK-32186][DOCS][PYTHON] Development - Debugging

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to document the way of debugging PySpark. It's pretty much self-descriptive.

I made a demo site to review it more effectively: https://hyukjin-spark.readthedocs.io/en/stable/development/debugging.html

### Why are the changes needed?

To let users know how to debug PySpark applications.

### Does this PR introduce _any_ user-facing change?

Yes, it adds a new page in the documentation about debugging PySpark.

### How was this patch tested?

Manually built the doc.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Again, I am going to proofread by myself few more times but .. cc @srowen, @viirya, @BryanCutler, @huaxingao, @zhengruifeng, @holdenk, @Fokko and @zero323. It would be great if I can have some feedback about it.

**Comment by HyukjinKwon** [comment]
FWIW, documenting such setup in a language dedicated page isn't a radical approach. We're already doing it for RStudio and SparkR at https://spark.apache.org/docs/latest/sparkr.html#starting-up-from-rstudio

File: python/docs/source/development/debugging.rst [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+=================
+Debugging PySpark
+=================
+
+PySpark uses Spark as an engine. In case that PySpark applications do not require the interaction
```
Comment (srowen) [inline python/docs/source/development/debugging.rst:22]: Slightly better is "If a PySpark application does not require interaction ..."

File: python/docs/source/development/debugging.rst [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+This page describes how to debug such Python workers instead of focusing on debugging with JVM.
+Profiling and debugging JVM is described at `Useful Developer Tools <https://spark.apache.org/developer-tools.html>`_.
+
+
+Remote Debugging (PyCharm)
+--------------------------
+
+In order to debug the Python workers remotely, you should connect from the Python worker to the debug server in PyCharm.
+In this section, it describes the remote debug within single machine to demonstrate easily.
```
Comment (srowen) [inline python/docs/source/development/debugging.rst:35]: Maybe: "This section describes remote debugging within a single machine .."

File: python/docs/source/development/debugging.rst [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+In order to debug PySpark applications in other machines, please refer to the full instructions that are specific
+to PyCharm is documented at `here <https://www.jetbrains.com/help/pycharm/remote-debugging-with-product.html#remote-debug-config>`_.
```
Comment (srowen) [inline python/docs/source/development/debugging.rst:37]: PyCharm, documented here

File: python/docs/source/development/debugging.rst [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+Remote Debugging (PyCharm)
+--------------------------
+
+In order to debug the Python workers remotely, you should connect from the Python worker to the debug server in PyCharm.
+In this section, it describes the remote debug within single machine to demonstrate easily.
+In order to debug PySpark applications in other machines, please refer to the full instructions that are specific
+to PyCharm is documented at `here <https://www.jetbrains.com/help/pycharm/remote-debugging-with-product.html#remote-debug-config>`_. 
+
+Firstly, choose **Run | Edit Configuration...** in the main manu, and it opens the Run/debug configurations dialog.
```
Comment (srowen) [inline python/docs/source/development/debugging.rst:39]: Choose Edit Configuration from the Run menu
manu -> menu

Start a new sentence. "It opens the Run/Debug ..."

File: python/docs/source/development/debugging.rst [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+You have to click ``+`` configuration on the toolbar, and from the list of available configurations, select **Python Debug Server**.
+Enter the name of this run/debug configuration, for example, ``MyRemoteDebugger`` and also specify the port number, for example ``12345``.
+
+.. image:: ../../../../docs/img/pyspark-remote-debug1.png
+    :alt: PyCharm remote debugger setting
+
+After that, you should install the corresponding version of ``pydevd-pycahrm`` package. In the previous dialog, it shows the command
```
Comment (srowen) [inline python/docs/source/development/debugging.rst:46]: of the ... package

File: python/docs/source/development/debugging.rst [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        #======================Copy and paste from the previous dialog===========================
+        import pydevd_pycharm
+        pydevd_pycharm.settrace('localhost', port=12345, stdoutToServer=True, stderrToServer=True)
+        #========================================================================================
+        worker.main(*args, **kwargs)
+    daemon.worker_main = remote_debug_wrapped
+    if __name__ == '__main__':
+        daemon.manager()" > remote_debug.py
+
+You will use this file as the Python workers in your PySpark applications by using ``spark.python.daemon.module`` configuration.
```
Comment (srowen) [inline python/docs/source/development/debugging.rst:68]: workers -> worker
using the ... configuration

File: python/docs/source/development/debugging.rst [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+Run ``pyspark`` shell with the configuration below:
```
Comment (srowen) [inline python/docs/source/development/debugging.rst:69]: Run the pyspark shell with ... , or Run pyspark with

File: python/docs/source/development/debugging.rst [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+.. code-block:: bash
+
+    pyspark --conf spark.python.daemon.module=remote_debug
+
+Now you're ready to remote debug. Start debug with your ``MyRemoteDebugger``.
```
Comment (srowen) [inline python/docs/source/development/debugging.rst:75]: Start debugging

File: python/docs/source/development/debugging.rst [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+.. image:: ../../../../docs/img/pyspark-remote-debug2.png
+    :alt: PyCharm run remote debugger
+
+After that, run a job that creates a Python workers, for example, as below:
```
Comment (srowen) [inline python/docs/source/development/debugging.rst:80]: creates Python workers

File: python/docs/source/development/debugging.rst [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+.. code-block:: python
+
+    spark.range(10).repartition(1).rdd.map(lambda x: x).collect()
+
+
+Checking Memory and CPU Usage
+-----------------------------
+
+Python workers are typically monitored via ``top`` and ``ps`` commands because Python workers launch multiple Python
```
Comment (srowen) [inline python/docs/source/development/debugging.rst:90]: Python workers create multiple Python processes

**Review by srowen — CHANGES_REQUESTED** [review]

File: python/docs/source/development/debugging.rst [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        daemon.manager()" > remote_debug.py
+
+You will use this file as the Python worker in your PySpark applications by using the ``spark.python.daemon.module`` configuration.
+Run the ``pyspark`` shell with the configuration below:
+
+.. code-block:: bash
+
+    pyspark --conf spark.python.daemon.module=remote_debug
+
+Now you're ready to remote debug. Start debugging with your ``MyRemoteDebugger``.
```
Comment (viirya) [inline python/docs/source/development/debugging.rst:74]: remote debug -> remotely debug

File: python/docs/source/development/debugging.rst [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+.. code-block:: python
+
+    spark.range(10).repartition(1).rdd.map(lambda x: x).collect()
+
+
+Checking Memory and CPU Usage
+-----------------------------
+
+Python workers are typically monitored via ``top`` and ``ps`` commands because Python workers create multiple Python processes
+workers are created as processes. As an example, you can ``ps`` as below:
```
Comment (viirya) [inline python/docs/source/development/debugging.rst:90]: Is `workers are created as processes` redundant?

**Review by viirya — APPROVED** [review]
Two minor comments. Otherwise LGTM.

**Comment by HyukjinKwon** [comment]
Thanks @srowen and @viirya.

File: python/docs/source/development/debugging.rst [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+PySpark uses Spark as an engine. If a PySpark application does not require interaction
+between Python workers and JVMs, Python workers are not launched. They are lazily launched only when
+Python native functions or data have to be handled, for example, when you execute pandas UDFs or
+PySpark RDD APIs.
+
+This page describes how to debug such Python applications and workers instead of focusing on debugging with JVM.
+Profiling and debugging JVM is described at `Useful Developer Tools <https://spark.apache.org/developer-tools.html>`_.
+
+
+Remote Debugging (PyCharm)
```
Comment (zero323) [inline python/docs/source/development/debugging.rst:31]: Can we explicitly state that the remote debugger is available only PyCharm Professional (I know, it is clearly explained in PyCharm docs, but not everyone will get there)? And maybe link to [pydev remote debugger docs](https://www.pydev.org/manual_adv_remote_debugger.html) as an alternative.

**Review by Fokko — APPROVED** [review]

File: python/docs/source/development/debugging.rst (hunk 13 above)
Comment (HyukjinKwon) [reply python/docs/source/development/debugging.rst:31]: ahhhh .. I didn't know it's only for PyCharm Professional ..

Comment (HyukjinKwon) [reply python/docs/source/development/debugging.rst:31]: > And maybe link to pydev remote debugger docs as an alternative.

This one, I think we'll leave it as a future work. It'd be nice if somebody document this as well.

**Comment by HyukjinKwon** [comment]
I'll leave this open for a couple of more days in case there are some more comments.

**Review by zero323 — APPROVED** [review]

File: python/docs/source/development/debugging.rst [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+
+Remote Debugging (PyCharm Professional)
+---------------------------------------
+
+In order to debug the Python workers remotely, you should connect from the Python worker to the debug server in PyCharm.
+This section describes remote debugging within a single machine to demonstrate easily.
+In order to debug PySpark applications on other machines, please refer to the full instructions that are specific
+to PyCharm, documented `here <https://www.jetbrains.com/help/pycharm/remote-debugging-with-product.html>`_.
+
+Firstly, choose **Edit Configuration...** from the **Run** menu. It opens the Run/debug Configurations dialog.
```
Comment (gengliangwang) [inline python/docs/source/development/debugging.rst:39]: Nit: **Run/Debug Configurations**

File: python/docs/source/development/debugging.rst [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+You have to click ``+`` configuration on the toolbar, and from the list of available configurations, select **Python Debug Server**.
+Enter the name of this run/debug configuration, for example, ``MyRemoteDebugger`` and also specify the port number, for example ``12345``.
```
Comment (gengliangwang) [inline python/docs/source/development/debugging.rst:41]: Nit: **Run/Debug Configurations**

**Review by BryanCutler — APPROVED** [review]
This will be great to have, I've gotten many questions from people on how to debug PySpark. Just a couple of points you might want to include:

- This is one approach to debug the workers, but there are other ways to debug also
- If debugging driver code only, you don't need to setup remote debugging

File: python/docs/source/development/debugging.rst [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+| After that, submit your application. This will connect to your PyCharm debugging server and enable you to debug on driver side remotely.
+
+.. code-block:: bash
+
+    spark-submit app.py
+
+Executor Side
+~~~~~~~~~~~~~
+
+To debug on the executor side, prepare a Python file as below In your current working directory.
```
Comment (viirya) [inline python/docs/source/development/debugging.rst:94]: In -> in

**Comment by HyukjinKwon** [comment]
Hey guys, sorry for a bit of noise here. Would you guys mind if I ask take another look? I changed a bit a lot while I am addressing @BryanCutler and @zero323 comments.

**Comment by HyukjinKwon** [comment]
The link here https://hyukjin-spark.readthedocs.io/en/stable/development/debugging.html is also updated.

File: python/docs/source/development/debugging.rst [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+a PySpark application does not require interaction between Python workers and JVMs. They are lazily launched only when
+Python native functions or data have to be handled, for example, when you execute pandas UDFs or
+PySpark RDD APIs.
+
+This page focuses on debugging Python side of PySpark on both driver and executor sides instead of focusing on debugging
+with JVM. Profiling and debugging JVM is described at `Useful Developer Tools <https://spark.apache.org/developer-tools.html>`_.
+
+Note that,
+
+- If you are running locally, you can directly debug the driver side via using your IDE without the remote debug feature.
```
Comment (HyukjinKwon) [inline python/docs/source/development/debugging.rst:38]: BTW, @itholic is working on documenting local PyCharm setup in another page (at SPARK-32189). We could add a link here once that page is finished.

**Review by zhengruifeng — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
I am going to merge this. I updated the link here: https://hyukjin-spark.readthedocs.io/en/stable/development/debugging.html 
Please feel free to leave a comment - I will address them in followups.

Merged to master.

**Comment by HyukjinKwon** [comment]
Thank you guys for reviewing this.

**Pull request closed without merging**